File: ts/src/message.ts
import { encryptDataRoomKey, decryptDataRoomKey, isValidRoomKey, isValidEncryptedDataRoomKey } from "./roomKey.ts";
import { isValidIdentityKeyPrivate, isValidIdentityKeyPublic, signIdentityKey, verifyIdentityKey } from "./identityKey.ts";
import type {
  Message,
  NotEncryptMessage,
  EncryptedMessage,
  NotEncryptMessageValue,
  TextContent,
  ImageContent,
  VideoContent,
  AudioContent,
  FileContent,
  ThumbnailContent,
  ReplyInfo,
  TextThumbnail,
  MediaThumbnail,
  FilesThumbnail
} from "./type.ts";
import { keyHash } from "./utils.ts";

/**
 * メッセージを暗号化する
 * @param message メッセージオブジェクト (NotEncryptMessageValue)
 * @param metadata メッセージメタデータ
 * @param roomKey ルームキー
 * @param identityKey アイデンティティキー情報
 * @param roomid ルームID
 * @returns 暗号化されたメッセージと署名、またはエラー時はnull
 */
export async function encryptMessage(
  messageValue: NotEncryptMessageValue,
  metadata: {
    channel: string;
    timestamp: number;
    isLarge: boolean;
    original?: string;
  },
  roomKey: string,
  identityKey: {
    privateKey: string;
    pubKeyHash: string;
  },
  roomid: string,
): Promise<
  {
    message: string;
    sign: string;
  } | null
> {
  if (!isValidRoomKey(roomKey)) {
    console.error("無効なルームキーが指定されました");
    return null;
  }
  if (!isValidIdentityKeyPrivate(identityKey.privateKey)) {
    console.error("無効なアイデンティティキーが指定されました");
    return null;
  }

  try {
    const messageContent = JSON.stringify(messageValue);

    const encryptedValue = await encryptDataRoomKey(roomKey, messageContent);
    if (!encryptedValue) {
      console.error("メッセージコンテンツの暗号化に失敗しました");
      return null;
    }

    const messageObj: EncryptedMessage = {
      encrypted: true,
      value: encryptedValue,
      channel: metadata.channel,
      timestamp: metadata.timestamp,
      isLarge: metadata.isLarge,
      original: metadata.original,
      roomid: roomid,
    };

    const messageString = JSON.stringify(messageObj);

    const sign = signIdentityKey(
      identityKey.privateKey,
      messageString,
      identityKey.pubKeyHash,
    );
    if (!sign) {
      console.error("メッセージの署名に失敗しました");
      return null;
    }

    return {
      message: messageString,
      sign: sign,
    };
  } catch (error) {
    console.error("メッセージ暗号化中にエラー:", error);
    return null;
  }
}

/**
 * 暗号化されたメッセージを復号する
 * @param message 暗号化されたメッセージと署名
 * @param serverData サーバーデータ（タイムスタンプ）
 * @param roomKey ルームキー
 * @param identityKeyPub アイデンティティキーの公開鍵
 * @param roomid ルームID
 * @returns 復号されたメッセージ (Message 型)、またはエラー時はnull
 */
export async function decryptMessage(
  message: {
    message: string;
    sign: string;
  },
  serverData: {
    timestamp: number;
  },
  roomKey: string,
  identityKeyPub: string,
  roomid: string,
): Promise<Message | null> {
  if (!isValidIdentityKeyPublic(identityKeyPub)) {
    console.error("無効なアイデンティティキーが指定されました");
    return null;
  }

  try {
    const { message: messageString, sign: signString } = message;
    const messageObj = JSON.parse(messageString) as Message;

    if (!verifyIdentityKey(identityKeyPub, signString, messageString)) {
      console.error("メッセージの署名検証に失敗しました");
      return null;
    }

    if (!messageObj.encrypted) {
      if (roomid !== messageObj.roomid) {
        console.error("メッセージのルームIDが一致しません");
        return null;
      }
      if (Math.abs(messageObj.timestamp - serverData.timestamp) > 60000) {
        console.error("メッセージのタイムスタンプが無効です");
        return null;
      }
      return messageObj as NotEncryptMessage;
    }

    if (!isValidRoomKey(roomKey)) {
      console.error("無効なルームキーが指定されました (復号時)");
      return null;
    }

    if (Math.abs(messageObj.timestamp - serverData.timestamp) > 60000) {
      console.error("メッセージのタイムスタンプが無効です");
      return null;
    }

    if (roomid !== messageObj.roomid) {
      console.error("メッセージのルームIDが一致しません");
      return null;
    }

    if (!isValidEncryptedDataRoomKey(messageObj.value)) {
      console.error("メッセージ内の暗号化データが無効です");
      return null;
    }

    const decryptedContentString = await decryptDataRoomKey(roomKey, messageObj.value);
    if (!decryptedContentString) {
      console.error("メッセージの復号に失敗しました");
      return null;
    }

    const decryptedValue = JSON.parse(decryptedContentString) as NotEncryptMessageValue;

    const decryptedMessage: NotEncryptMessage = {
      encrypted: false,
      value: decryptedValue,
      channel: messageObj.channel,
      timestamp: messageObj.timestamp,
      isLarge: messageObj.isLarge,
      original: messageObj.original,
      roomid: messageObj.roomid
    };

    return decryptedMessage;

  } catch (error) {
    console.error("メッセージ復号中にエラー:", error);
    return null;
  }
}

/**
 * メッセージが有効かどうかを検証する
 * @param message 検証するメッセージ文字列 (Message 型の JSON 文字列)
 * @returns 検証結果（true: 有効、false: 無効）
 */
export function isValidMessage(message: string): boolean {
  try {
    const messageObj = JSON.parse(message) as Message;

    if (typeof messageObj.encrypted !== 'boolean') return false;
    if (typeof messageObj.channel !== 'string' || messageObj.channel.length > 100) return false;
    if (typeof messageObj.timestamp !== 'number' || !Number.isInteger(messageObj.timestamp)) return false;
    if (typeof messageObj.isLarge !== 'boolean') return false;
    if (messageObj.original !== undefined && typeof messageObj.original !== 'string') return false;
    if (typeof messageObj.roomid !== 'string') return false;

    if (messageObj.encrypted) {
      if (typeof messageObj.value !== 'string') return false;
      if (!isValidEncryptedDataRoomKey(messageObj.value)) {
        console.error("暗号化メッセージの value が無効です");
        return false;
      }
      return true;
    } else {
      const value = messageObj.value;
      if (typeof value !== 'object' || value === null) return false;

      const validTypes = ["text", "image", "video", "audio", "file", "thumbnail"];
      if (!validTypes.includes(value.type)) {
        console.error("メッセージタイプが無効です");
        return false;
      }

      if (typeof value.content !== 'string') return false;

      if (value.reply !== undefined) {
        if (typeof value.reply !== 'object' || value.reply === null || typeof value.reply.id !== 'string') {
          console.error("reply が無効です");
          return false;
        }
      }

      if (value.mention !== undefined) {
        if (!Array.isArray(value.mention) || !value.mention.every(m => typeof m === 'string')) {
          console.error("mention が無効です");
          return false;
        }
      }

      try {
        JSON.parse(value.content);
      } catch {
        console.error("content が有効な JSON 文字列ではありません");
        return false;
      }

      return true;
    }
  } catch (error) {
    console.error("メッセージ検証中にエラー:", error);
    return false;
  }
}

/**
 * テキストコンテンツを作成する
 * @param text テキスト
 * @param format フォーマット（省略時はplain）
 * @param options オプション (isThumbnail, thumbnailOf, originalSize)
 * @returns テキストコンテンツオブジェクトのJSON文字列
 */
export function createTextContent(
  text: string,
  format: "plain" | "markdown" = "plain",
  options: { isThumbnail?: boolean; thumbnailOf?: string; originalSize?: number } = {}
): string {
  const content: TextContent = {
    text,
    format,
    ...options
  };
  return JSON.stringify(content);
}

/**
 * 画像コンテンツを作成する
 * @param uri 画像URI（Base64またはURL）
 * @param filename ファイル名
 * @param mimeType MIMEタイプ
 * @param options オプション (isThumbnail, thumbnailOf, originalSize)
 * @returns 画像コンテンツオブジェクトのJSON文字列
 */
export function createImageContent(
  uri: string,
  filename: string,
  mimeType: string,
  options: { isThumbnail?: boolean; thumbnailOf?: string; originalSize?: number } = {}
): string {
  const content: ImageContent = {
    uri,
    metadata: {
      filename,
      mimeType
    },
    ...options
  };
  return JSON.stringify(content);
}

/**
 * 動画コンテンツを作成する
 * @param uri 動画URI（Base64またはURL）
 * @param filename ファイル名
 * @param mimeType MIMEタイプ
 * @param options オプション (isThumbnail, thumbnailOf, originalSize)
 * @returns 動画コンテンツオブジェクトのJSON文字列
 */
export function createVideoContent(
  uri: string,
  filename: string,
  mimeType: string,
  options: { isThumbnail?: boolean; thumbnailOf?: string; originalSize?: number } = {}
): string {
  const content: VideoContent = {
    uri,
    metadata: {
      filename,
      mimeType
    },
    ...options
  };
  return JSON.stringify(content);
}

/**
 * 音声コンテンツを作成する
 * @param uri 音声URI（Base64またはURL）
 * @param filename ファイル名
 * @param mimeType MIMEタイプ
 * @param options オプション (isThumbnail, thumbnailOf, originalSize)
 * @returns 音声コンテンツオブジェクトのJSON文字列
 */
export function createAudioContent(
  uri: string,
  filename: string,
  mimeType: string,
  options: { isThumbnail?: boolean; thumbnailOf?: string; originalSize?: number } = {}
): string {
  const content: AudioContent = {
    uri,
    metadata: {
      filename,
      mimeType
    },
    ...options
  };
  return JSON.stringify(content);
}

/**
 * ファイルコンテンツを作成する
 * @param uri ファイルURI（Base64またはURL）
 * @param filename ファイル名
 * @param mimeType MIMEタイプ
 * @param options オプション (isThumbnail, thumbnailOf, originalSize)
 * @returns ファイルコンテンツオブジェクトのJSON文字列
 */
export function createFileContent(
  uri: string,
  filename: string,
  mimeType: string,
  options: { isThumbnail?: boolean; thumbnailOf?: string; originalSize?: number } = {}
): string {
  const content: FileContent = {
    uri,
    metadata: {
      filename,
      mimeType
    },
    ...options
  };
  return JSON.stringify(content);
}

File: rust/src/message.rs
use crate::r#type::{NotEncryptMessageValue, NotEncryptMessage, EncryptedMessage, TextContent, ImageContent};
use crate::schema::validate_message;
use crate::room_key::{encrypt_data_room_key, decrypt_data_room_key, is_valid_room_key, is_valid_encrypted_data_room_key};
use crate::identity_key::{is_valid_identity_key_private, is_valid_identity_key_public, sign_identity_key, verify_identity_key};
use crate::account_key::encrypt_data_account_key;
use serde_json::{Value, json};

pub fn encrypt_message(
    message_value_json: &str,
    metadata_json: &str,
    room_key_json: &str,
    identity_priv_json: &str,
    identity_pubhash: &str,
    roomid: &str,
) -> Option<String> {
    if !is_valid_room_key(room_key_json) { return None; }
    if !is_valid_identity_key_private(identity_priv_json) { return None; }
    println!("debug1");
    let encrypted_val = encrypt_data_room_key(room_key_json, message_value_json)?;
    let meta: Value = serde_json::from_str(metadata_json).ok()?;
    let channel = meta.get("channel")?.as_str()?.to_string();
    let timestamp = meta.get("timestamp")?.as_u64()?;
    let is_large = meta.get("isLarge")?.as_bool()?;
    let original = meta.get("original").and_then(|v| v.as_str()).map(String::from);
    let msg = EncryptedMessage { encrypted: true, value: encrypted_val.clone(), channel: channel.clone(), original: original.clone(), timestamp, is_large, roomid: roomid.to_string() };
    let msg_str = serde_json::to_string(&msg).ok()?;
    let sign = sign_identity_key(identity_priv_json, &msg_str, identity_pubhash)?;
    let res = json!({"message": msg_str, "sign": sign});
    serde_json::to_string(&res).ok()
}

pub fn decrypt_message(
    message_str: &str,
    sign_str: &str,
    server_timestamp: u64,
    room_key_json: &str,
    identity_pub_json: &str,
    roomid: &str,
) -> Option<String> {
    if !is_valid_identity_key_public(identity_pub_json) { return None; }
    if !verify_identity_key(identity_pub_json, sign_str, message_str) { return None; }
    let v: Value = serde_json::from_str(message_str).ok()?;
    let encrypted = v.get("encrypted")?.as_bool()?;
    let timestamp = v.get("timestamp")?.as_u64()?;
    let channel = v.get("channel")?.as_str()?.to_string();
    let is_large = v.get("isLarge")?.as_bool()?;
    let original = v.get("original").and_then(|v| v.as_str()).map(String::from);
    let rid = v.get("roomid")?.as_str()?;
    if rid != roomid || (timestamp as i64 - server_timestamp as i64).abs() as u64 > 60000 { return None; }
    if !encrypted {
        let val_json = v.get("value")?.clone();
        let res = json!({
            "encrypted": false,
            "value": val_json,
            "channel": channel,
            "original": original,
            "timestamp": timestamp,
            "isLarge": is_large,
            "roomid": roomid
        });
        return serde_json::to_string(&res).ok();
    }
    if !is_valid_room_key(room_key_json) { return None; }
    let enc_val = v.get("value")?.as_str()?;
    if !is_valid_encrypted_data_room_key(enc_val) { return None; }
    let decrypted_str = decrypt_data_room_key(room_key_json, enc_val)?;
    let val_json: Value = serde_json::from_str(&decrypted_str).ok()?;
    // Wrap decrypted content into NotEncryptMessageValue struct
    let content_type = if val_json.get("text").is_some() {
        "text"
    } else if val_json.get("uri").is_some() {
        "image"
    } else {
        "text"
    };
    let value_obj = json!({
        "type": content_type,
        "content": decrypted_str
    });
    let res = json!({
        "encrypted": false,
        "value": value_obj,
        "channel": channel,
        "original": original,
        "timestamp": timestamp,
        "isLarge": is_large,
        "roomid": roomid
    });
    serde_json::to_string(&res).ok()
}

pub fn is_valid_message(message_str: &str) -> bool {
    serde_json::from_str::<Value>(message_str)
        .ok()
        .map_or(false, |v| validate_message(&v))
}

pub fn create_text_content(
    text: &str,
    format: Option<&str>,
    is_thumbnail: Option<bool>,
    thumbnail_of: Option<&str>,
    original_size: Option<u64>,
) -> Option<String> {
    let content = TextContent { text: text.to_string(), format: format.map(String::from), is_thumbnail, thumbnail_of: thumbnail_of.map(String::from), original_size };
    serde_json::to_string(&content).ok()
}

pub fn create_image_content(
    uri: &str,
    filename: &str,
    mime_type: &str,
    is_thumbnail: Option<bool>,
    thumbnail_of: Option<&str>,
    original_size: Option<u64>,
) -> Option<String> {
    let metadata = crate::r#type::MediaMetadata { filename: filename.to_string(), mime_type: mime_type.to_string() };
    let content = ImageContent { uri: uri.to_string(), metadata, is_thumbnail, thumbnail_of: thumbnail_of.map(String::from), original_size };
    serde_json::to_string(&content).ok()
}

pub fn create_video_content(
    uri: &str,
    filename: &str,
    mime_type: &str,
    is_thumbnail: Option<bool>,
    thumbnail_of: Option<&str>,
    original_size: Option<u64>,
) -> Option<String> {
    create_image_content(uri, filename, mime_type, is_thumbnail, thumbnail_of, original_size)
}

pub fn create_audio_content(
    uri: &str,
    filename: &str,
    mime_type: &str,
    is_thumbnail: Option<bool>,
    thumbnail_of: Option<&str>,
    original_size: Option<u64>,
) -> Option<String> {
    create_image_content(uri, filename, mime_type, is_thumbnail, thumbnail_of, original_size)
}

pub fn create_file_content(
    uri: &str,
    filename: &str,
    mime_type: &str,
    is_thumbnail: Option<bool>,
    thumbnail_of: Option<&str>,
    original_size: Option<u64>,
) -> Option<String> {
    create_image_content(uri, filename, mime_type, is_thumbnail, thumbnail_of, original_size)
}

pub fn encrypt_room_key_with_account_keys(
    users_json: &str,
    room_key_json: &str,
) -> Option<String> {
    let users: Vec<Value> = serde_json::from_str(users_json).ok()?;
    let mut res = Vec::new();
    for u in users {
        let account_key = u.get("accountKey")?.as_str()?;
        let user_id = u.get("userId")?.as_str()?;
        if let Some(enc) = encrypt_data_account_key(account_key, room_key_json) {
            res.push(json!({"userId": user_id, "encryptedData": enc}));
        }
    }
    serde_json::to_string(&res).ok()
}